libs/vkd3d/swapchain.c

```diff
@@ -2686,9 +2703,71 @@ static HRESULT STDMETHODCALLTYPE d3d12_swapchain_ResizeBuffers1(dxgi_swapchain_i
     return hr;
 }
 
+static void d3d12_swapchain_update_hdr_metadata(struct d3d12_swapchain *swapchain, DXGI_HDR_METADATA_TYPE metadata_type, VkHdrMetadataEXT *metadata)
+{
+    if (swapchain->hdr_metadata_type == metadata_type)
+        return;
+
+    swapchain->hdr_metadata_type = metadata_type;
+
+    if (!swapchain->command_queue->device->vk_info.EXT_hdr_metadata)
+        return;
+
+    if (metadata)
+    {
+        assert(metadata_type == DXGI_HDR_METADATA_TYPE_HDR10);
+
+        swapchain->hdr_metadata = *metadata;
+        d3d12_swapchain_set_hdr_metadata(swapchain);
+    }
+    else
+    {
+        assert(metadata_type == DXGI_HDR_METADATA_TYPE_NONE);
+
+        /* If we are DXGI_HDR_METADATA_TYPE_NONE, we must clear the metadata.
+         * The D3D docs do not mention this, but the sample does!
+         * We have no way of doing this easily in Vulkan, aside from just
+         * re-creating the swapchain. */
+        d3d12_swapchain_destroy_resources(swapchain, false);
+        d3d12_swapchain_create_vulkan_swapchain(swapchain, false);
+    }
+}
+
+static VkXYColorEXT convert_xy_color(UINT16 *dxgi_color)
+{
+    return (VkXYColorEXT){ dxgi_color[0] / 50000.0f, dxgi_color[1] / 50000.0f };
+}
+
+static float convert_luminance(UINT dxgi_luminance)
+{
+    return dxgi_luminance / 0.0001f;
+}
+
+static float convert_level(UINT16 dxgi_level)
+{
+    return (float)dxgi_level;
+}
+
+static VkHdrMetadataEXT convert_hdr_metadata_hdr10(DXGI_HDR_METADATA_HDR10 *dxgi_metadata)
+{
+    VkHdrMetadataEXT vulkan_metadata = { VK_STRUCTURE_TYPE_HDR_METADATA_EXT };
+    vulkan_metadata.displayPrimaryRed = convert_xy_color(dxgi_metadata->RedPrimary);
+    vulkan_metadata.displayPrimaryGreen = convert_xy_color(dxgi_metadata->GreenPrimary);
+    vulkan_metadata.displayPrimaryBlue = convert_xy_color(dxgi_metadata->BluePrimary);
+    vulkan_metadata.whitePoint = convert_xy_color(dxgi_metadata->WhitePoint);
+    vulkan_metadata.maxLuminance = convert_luminance(dxgi_metadata->MaxMasteringLuminance);
```

Comment:
Is this correct? Docs say:

MaxMasteringLuminance

The maximum number of nits of the display used to master the content. Values are in **whole** nits.

MinMasteringLuminance

The minimum number of nits of the display used to master the content. Values are 1/10000th of a nit (0.0001 nit).

Reply:
The HDR samples does not treat MaxMasteringLuminance as whole nits...

https://github.com/microsoft/DirectX-Graphics-Samples/blob/master/Samples/Desktop/D3D12HDR/src/D3D12HDR.cpp#L994

I will investigate further.

Reply:
https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_5/ns-dxgi1_5-dxgi_hdr_metadata_hdr10#remarks suggests it's as the docs say. Maybe this struct is inherited directly from HEVC specs ... The data types are so silly it has to be a video spec :frog:

Reply:
```This structure represents the definition of HDR10 metadata used with HEVC, not HDR10 metadata for ST.2086. These are closely related but defined differently.```


HEVC spec states:

```
max_display_mastering_luminance and min_display_mastering_luminance specify the nominal maximum and
minimum display luminance, respectively, of the mastering display in units of 0.0001 candelas per square metre.
min_display_mastering_luminance shall be less than max_display_mastering_luminance.
```

Which seems to indicate the docs may be wrong here.

Reply:
I made it just follow the docs for now and added a comment.